### PR TITLE
 sock: amend API to iterate over stack-internal buffer chunks

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
With lwIP we have a chunked UDP payload, so just providing the stack-internal buffer is not possible. To be able to iterate over such a chunked payload, this change allows the `sock_*_recv_buf()` functions to use the internal buffer context as an iteration state. Since the newly required handling of the context requires an iteration anyways, `sock_recv_buf_free()` becomes unnecessary.

As this function is still experimental (I piggy-backed the badge for that) this API change only should require 1 ACK.

### Testing procedure
Read the doc and

```sh
for test in tests/gnrc_sock_{ip,udp}; do
    make -C ${test} flash test
done
```

should still pass.

### Issues/PRs references
Needed for #13701.

### CODE REVIEW

**Comment by benpicco** [comment]
> Since GNRC does not chunk the UDP payload, an adaptation for this API change is not required.

That's not entirely true - it must now return 0 when called with `*buf_ctx != NULL`

**Comment by benpicco** [comment]
I think the API would be easier to use if instead you introduce an additional `bool *more` parameter.
That way the user doesn't have to call the function twice to find out if there is more data.

**Comment by miri64** [comment]
> I think the API would be easier to use if instead you introduce an additional `bool *more` parameter.
> That way the user doesn't have to call the function twice to find out if there is more data.

I'd like to keep the API minimal. The correct way to use it would be in a while-loop anyway.

**Comment by benpicco** [comment]
Just noticed - you should probably also amend `tests/gnrc_sock_udp/main.c`

**Comment by miri64** [comment]
I just had another idea :sweat_smile:

**Comment by miri64** [comment]
Ok, I restructured it a bit and now it somewhat addresses your comment about having to call this function more than once. This is still the case, however, the second call will release the internal context implicitly (what was required to before explicitly using `sock_recv_buf_free()`. So you still need an extra call, but do not require to call `sock_recv_buf_free()` in the end :-).

**Comment by benpicco** [comment]
Nice. That 'forces' the user to always use the API properly and recuses the API surface.

I reckon there is no use for the content of `buf_ctx` outside that function?

**Comment by benpicco** [comment]
(Murdock has found some issues)

**Comment by miri64** [comment]
> I reckon there is no use for the content of `buf_ctx` outside that function?

Mostly yes. For the lwIP implementation I use the `struct netconn` directly to determine the total length for `-ENOBUFS`, but that stays internal to the lwIP wrapper implementation.

**Comment by miri64** [comment]
> (Murdock has found some issues)

Huh? The change did not run yet.

**Comment by benpicco** [comment]
Oh sorry, I confused this with #13701

File: sys/net/gnrc/sock/ip/gnrc_sock_ip.c [hunk 1]
```diff
@@ -90,20 +90,21 @@ ssize_t sock_ip_recv(sock_ip_t *sock, void *data, size_t max_len,
...
+    bool nobufs = false;
 
     assert((sock != NULL) && (data != NULL) && (max_len > 0));
-    res = sock_ip_recv_buf(sock, &pkt, &ctx, timeout, remote);
-    if (res >= 0) {
+    while ((res = sock_ip_recv_buf(sock, &pkt, &ctx, timeout, remote)) > 0) {
         if (res > (ssize_t)max_len) {
-            res = -ENOBUFS;
+            nobufs = true;
+            continue;
```
Comment (benpicco) [inline sys/net/gnrc/sock/ip/gnrc_sock_ip.c:101]: `break`?
Otherwise this will loop forever.

Comment (benpicco) [reply sys/net/gnrc/sock/ip/gnrc_sock_ip.c:101]: Ah never mind - the next iteration will free `ctx`

Comment (miri64) [reply sys/net/gnrc/sock/ip/gnrc_sock_ip.c:101]: No, we want to have another call to `sock_ip_recv_buf` to free the buffer space ;-).

**Comment by benpicco** [comment]
This looks good now - please squash.

**Comment by miri64** [comment]
> This looks good now - please squash.

Done. Also fixed some doc issues while I was at it ;-).

**Pull request merged**